### PR TITLE
Enable flook

### PR DESCRIPTION
**Pull request by pfebrer**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by zerothi** [comment]
build-number update required, otherwise we can check it.

**Comment by pfebrer** [comment]
> build-number update required, otherwise we can check it.

You are too fast, I was doing it :)

**Comment by pfebrer** [comment]
@conda-forge-admin, please rerender

**Comment by pfebrer** [comment]
Is there some way to try with the current master of siesta? Or is it only possible with tagged releases?

**Comment by pfebrer** [comment]
The MAC build seems to have gone over flook without problems, only the linux one has failed

**Comment by zerothi** [comment]
> Is there some way to try with the current master of siesta? Or is it only possible with tagged releases?

conda releases should stick to a formal release, IMHO. So I think you have these options:
1) wait until rc2 (or 5.0.0)
2) try and fix the flook compilation by adding a patch which addresses only this option.

From the looks of it, then it seems the problem will be present in later commits as well. It could be env-vars. Play with that.

**Comment by pfebrer** [comment]
Yes, I just wanted to try here. What you build here in the PR doesn't get published, does it?

**Comment by zerothi** [comment]
No, only once merged. :)

**Comment by pfebrer** [comment]
@conda-forge-admin, please rerender

**Comment by pfebrer** [comment]
Not pretty, but it works :)

File: recipe/build.sh
Comment (zerothi) [inline recipe/build.sh:68]: I don't like this. Very non-standard.

I would suggest you fix the `setup.make` input file instead, via a patch.

This should also ensure that readline is captured correctly.

File: recipe/meta.yaml
Comment (zerothi) [inline recipe/meta.yaml:47]: Readline shouldn't be in `build` section.

**Review by zerothi — CHANGES_REQUESTED** [review]
Yeah, it is too hacky, could you try with fixing the setup.make instead?

**Comment by pfebrer** [comment]
I have the impression that `CMake_C_Compiler` is simply not set. In MAC, a `gcc` binary just happens to be  in the path, that's why it works.

https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=832491&view=logs&j=65769002-1136-5bbb-5e91-b3fbb7cdd70e&t=fdb86174-6733-5ed6-1dd5-f2594c29f582&l=2456

The CMake_*_compiler are defined here:

https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=832491&view=logs&j=65769002-1136-5bbb-5e91-b3fbb7cdd70e&t=fdb86174-6733-5ed6-1dd5-f2594c29f582&l=491

But only the fortran ones are set. I looked through the CMakeLists file and I can't find where `Compilers` is requested. Do you know where this is? The right fix would be then to also ask for the C compiler to be defined.

**Comment by pfebrer** [comment]
Isn't that equally hacky? I think this is a problem for any SIESTA compilation, not just the conda one. It only works if there is by chance a `gcc` in the path. Perhaps I should submit an issue on the siesta repo instead.

**Comment by zerothi** [comment]
> Isn't that equally hacky? I think this is a problem for any SIESTA compilation, not just the conda one. It only works if there is by chance a `gcc` in the path. Perhaps I should submit an issue on the siesta repo instead.

No, not really. The patch should point the CC to the correct GCC variable. In this case it is how conda-forge setups its build procedure. And that may be very different from local installations.
Moving and copying is executables are much more hacky since they can use other linker lines depending on location (which can be dangerous).

I think this is not a problem in the siesta repo as the cmake invocation correctly sets the CC variable in `setup.make`, so IMHO it looks like a missing variable in conda.

**Comment by jan-janssen** [comment]
You can use:
```
cmake -S. -Bobj_cmake "${cmake_opts[@]}" ${CMAKE_ARGS} 
```
For more explanations https://conda-forge.org/docs/maintainer/knowledge_base.html#using-cmake

**Comment by zerothi** [comment]
> You can use:
> 
> ```
> cmake -S. -Bobj_cmake "${cmake_opts[@]}" ${CMAKE_ARGS} 
> ```
> 
> For more explanations https://conda-forge.org/docs/maintainer/knowledge_base.html#using-cmake
Ah, I always assumed CMAKE_ARGS was automatically read, then please try that here, yes! :)

**Comment by pfebrer** [comment]
Ok, so I found out that `CC` in `setup.make` is set to:

`$BUILD_PREFIX/bin/x86_64-conda-linux-gnu-cc`

which exists, and it is a symbolic link to 

`$BUILD_PREFIX/bin/x86_64-conda-linux-gnu-gcc`

So it is not a problem of environment variables. But somehow this is ignored and `gcc` is used instead.

**Comment by pfebrer** [comment]
I think flook just ignores whatever you set up as C compiler in `setup.make`.

I tried with:

```
PLATFORM=linux
CC = fake_cc
FC = fake_fc
#
.f90.o:
        $(FC) -c $(FFLAGS) $(INC) $<
.F90.o:
        $(FC) -c $(FFLAGS) $(INC) $<
.c.o:
        $(CC) -c $(CFLAGS) $(INC) $<
```

And I get:

```
setup.make:9: warning: overriding recipe for target '.f90.o'
setup.make:9: warning: ignoring old recipe for target '.f90.o'
setup.make:11: warning: overriding recipe for target '.F90.o'
setup.make:11: warning: ignoring old recipe for target '.F90.o'
setup.make:13: warning: overriding recipe for target '.c.o'
setup.make:13: warning: ignoring old recipe for target '.c.o'
(mkdir -p aotus/obj ; cd aotus/obj ; \
	echo "TOP_DIR =../.././aotus" > Makefile ; \
	echo "PLATFORM = linux" >> Makefile ; \
	echo "include ../.././aotus/Makefile.smeka" >> Makefile )
(cd aotus/obj ; make SETUP=../../setup.make )
make[1]: Entering directory '/home/pfebrer/Simulations/siesta/siesta-feedstock/build_artifacts/siesta_1701257065502/work/obj_cmake/External/Lua-Engine/flook-prefix/src/flook-build/aotus/obj'
mkdir -p external
cp -rf ../.././aotus/external/lua-5.3.5 external/
make -C external/lua-5.3.5 linux
make[2]: Entering directory '/home/pfebrer/Simulations/siesta/siesta-feedstock/build_artifacts/siesta_1701257065502/work/obj_cmake/External/Lua-Engine/flook-prefix/src/flook-build/aotus/obj/external/lua-5.3.5'
cd src && make linux
make[3]: Entering directory '/home/pfebrer/Simulations/siesta/siesta-feedstock/build_artifacts/siesta_1701257065502/work/obj_cmake/External/Lua-Engine/flook-prefix/src/flook-build/aotus/obj/external/lua-5.3.5/src'
make all SYSCFLAGS="-DLUA_USE_LINUX" SYSLIBS="-Wl,-E -ldl -lreadline"
make[4]: Entering directory '/home/pfebrer/Simulations/siesta/siesta-feedstock/build_artifacts/siesta_1701257065502/work/obj_cmake/External/Lua-Engine/flook-prefix/src/flook-build/aotus/obj/external/lua-5.3.5/src'
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o lapi.o lapi.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o lcode.o lcode.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o lctype.o lctype.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o ldebug.o ldebug.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o ldo.o ldo.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o ldump.o ldump.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o lfunc.o lfunc.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o lgc.o lgc.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o llex.o llex.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o lmem.o lmem.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o lobject.o lobject.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o lopcodes.o lopcodes.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o lparser.o lparser.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o lstate.o lstate.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o lstring.o lstring.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o ltable.o ltable.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o ltm.o ltm.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o lundump.o lundump.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o lvm.o lvm.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o lzio.o lzio.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o lauxlib.o lauxlib.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o lbaselib.o lbaselib.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o lbitlib.o lbitlib.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o lcorolib.o lcorolib.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o ldblib.o ldblib.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o liolib.o liolib.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o lmathlib.o lmathlib.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o loslib.o loslib.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o lstrlib.o lstrlib.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o ltablib.o ltablib.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o lutf8lib.o lutf8lib.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o loadlib.o loadlib.c
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o linit.o linit.c
ar rcu liblua.a lapi.o lcode.o lctype.o ldebug.o ldo.o ldump.o lfunc.o lgc.o llex.o lmem.o lobject.o lopcodes.o lparser.o lstate.o lstring.o ltable.o ltm.o lundump.o lvm.o lzio.o lauxlib.o lbaselib.o lbitlib.o lcorolib.o ldblib.o liolib.o lmathlib.o loslib.o lstrlib.o ltablib.o lutf8lib.o loadlib.o linit.o 
ar: `u' modifier ignored since `D' is the default (see `U')
ranlib liblua.a
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o lua.o lua.c
gcc -std=gnu99 -o lua   lua.o liblua.a -lm -Wl,-E -ldl -lreadline 
gcc -std=gnu99 -O2 -Wall -Wextra -DLUA_COMPAT_5_2 -DLUA_USE_LINUX    -c -o luac.o luac.c
gcc -std=gnu99 -o luac   luac.o liblua.a -lm -Wl,-E -ldl -lreadline 
make[4]: Leaving directory '/home/pfebrer/Simulations/siesta/siesta-feedstock/build_artifacts/siesta_1701257065502/work/obj_cmake/External/Lua-Engine/flook-prefix/src/flook-build/aotus/obj/external/lua-5.3.5/src'
make[3]: Leaving directory '/home/pfebrer/Simulations/siesta/siesta-feedstock/build_artifacts/siesta_1701257065502/work/obj_cmake/External/Lua-Engine/flook-prefix/src/flook-build/aotus/obj/external/lua-5.3.5/src'
make[2]: Leaving directory '/home/pfebrer/Simulations/siesta/siesta-feedstock/build_artifacts/siesta_1701257065502/work/obj_cmake/External/Lua-Engine/flook-prefix/src/flook-build/aotus/obj/external/lua-5.3.5'
ar -x external/lua-5.3.5/src/liblua.a
fake_fc  -c -o dump_lua_fif_module.o  -Iexternal/lua-5.3.5/src ../.././aotus/LuaFortran/dump_lua_fif_module.f90
make[1]: fake_fc: Command not found
make[1]: *** [../.././aotus/smeka/Makefile.compiler:141: dump_lua_fif_module.o] Error 127
make[1]: Leaving directory '/home/pfebrer/Simulations/siesta/siesta-feedstock/build_artifacts/siesta_1701257065502/work/obj_cmake/External/Lua-Engine/flook-prefix/src/flook-build/aotus/obj'
make: *** [Makefile:67: lib-aotus] Error 2
```

So it uses gcc to compile C files but then it tries to use `FC` for fortran, and it fails (as it should because `fake_fc` doesn't exist).

**Comment by pfebrer** [comment]
It also ignores the includes (I think it just ignores the recipe for `.c.o`), that's why `readline` headers are not found. 

So I think this is a problem that has to be solved in flook. Here it can only be solved in a hacky manner.

**Comment by zerothi** [comment]
Perhaps a way around this would be to:
```shell
cd flook/aotus/external/lua
make CC=$GCC linux
```
which should respect it, then a subsequent make shouldn't rebuild as it is already done.
Try out locally ;)

**Comment by pfebrer** [comment]
Shouldn't this be managed by flook? It will happen to all users trying to compile SIESTA, not just here in the conda compilation. Also the proposed workaround adds the extra requirement that lua must be pre checked out, it can't be downloaded on the fly by cmake during the compilation.

**Comment by pfebrer** [comment]
In flook's Makefile you could call Make on lua before calling it on aotus, no?

**Comment by zerothi** [comment]
> In flook's Makefile you could call Make on lua before calling it on aotus, no?

I think not (I could be wrong though).
I think it is called through aotus.

**Comment by pfebrer** [comment]
So then, there could be a bash script or something?

**Comment by zerothi** [comment]
I think it is a matter of just cd'ing into `aotus/lua` and build it manually, then when one does siesta installation, it won't run the compilation again.

**Comment by pfebrer** [comment]
I can work around the problem by just creating a symbolic link `gcc -> real_cc_compiler` and it is much less complicated :) 

If you agree that this is a bug and could be fixed in flook (and it seems like an easy fix), I don't see why we need to make things complicated (and waste time waiting for the very slow build feedback loop).

**Comment by pfebrer** [comment]
> I don't like doing symbolic links in conda-builds. I am worried it has side-effects we are not aware of.

The C compiler that cmake uses (and that should be passed to lua make) is already a symbolic link, `x86_64-conda-linux-gnu-cc -> x86_64-conda-linux-gnu-gcc`.

**Comment by pfebrer** [comment]
I'm afraid if I try to do it in the other way I will spend the whole day looking at logs and figuring out details with one iteration every 5 minutes, and it is very discouraging. But whatever, I can try it again when I have nothing to do.

**Comment by zerothi** [comment]
I think you have somewhat misunderstood what I have tried to convey...

I don't really have much time now, so if you could help with the below things it would be much simpler.

In flook-repo
1. create a PR that fixes this issue, it should be a minimal intervention if it touches the `aotus` subdirectory as that is copied directly from the `aotus` repo
2. Ensure it picks-up the `setup.make` variables

In the siesta-feedstock-repo
1. Apply the same method you did in the PR for flook, but in the `build.sh`, then you shouldn't be required to do any weird testing as it should have been tested.
Again, this can be tested locally as you can checkout the 0.8.1 release, and try a simple build script to see if it picks-up the correct variables. So all-in-all everything should be testable locally, at least to a large extent.

**Comment by pfebrer** [comment]
@conda-forge-admin, please rerender

**Comment by pfebrer** [comment]
Ok, so I have come up with a Makefile that works. However I don't think it should be used in general because it calls `make` on `aotus/external/lua*` instead of `aotus/obj/external/lua*`.

If I remove `readline` from `build`, the build fails because it can not find `readline/readline.h`.

**Comment by pfebrer** [comment]
This is very confusing, the only build that fails is `linux_64_mpinompi`, because it doesn't find `readline`'s headers. But if I build locally (using the `build_locally.py` script) it works. What's the difference between CI and building locally? They both use a Docker container, no?

I need to publish a tutorial for SIESTA using `lua`, and the best thing would be to just ask people to install from `conda`. If someone can help here that would be really great, I don't know what else to try :(

**Comment by pfebrer** [comment]
So finally I took the approach of requiring `lua` as a dependency so that we don't have to compile it. It avoids modifying anything in flook.

The only extra thing to do is to tell `cmake` to find and link the `lua` library when compiling SIESTA, because conda installs it as a dynamic library and therefore it can't be included in `libflookall` (or at least I don't know how).

As ou can see it is all green :green_circle: :heavy_check_mark:  So @albgar @jan-janssen @zerothi can any of you merge this?

File: recipe/build.sh
Comment (zerothi) [inline recipe/build.sh:74]: Could you change the `CMakeLists.txt` to a patch, instead of a full file?

**Review by zerothi — CHANGES_REQUESTED** [review]
Only thing that needs changing is that the cmakelists.txt should be a patch file, not a copy file.

File: recipe/build.sh
Comment (zerothi) [reply recipe/build.sh:74]: See on the 4.1 branch how to add a patch to the `meta.yaml` file.

**Comment by zerothi** [comment]
And remove the left over file. Then we should just wait for the green lights.

**Comment by pfebrer** [comment]
All green :green_circle:

**Comment by zerothi** [comment]
thanks!!!

**Comment by pfebrer** [comment]
Now it failed in main due to a timeout :sweat_smile:

**Comment by zerothi** [comment]
> Now it failed in main due to a timeout 😅

annoying, and I can't get it to rerun until all the others have finished, basically the tests for this branch also required two re-runs... :(

**Pull request merged**
